From 115eca1edbb42d64f53b7eadad75a0e5d6c056e8 Mon Sep 17 00:00:00 2001
From: Elena Popa <elena.popa@nxp.com>
Date: Mon, 17 Feb 2025 14:08:44 +0200
Subject: [PATCH] imx93: add falcon mode support

Signed-off-by: Elena Popa <elena.popa@nxp.com>
---
 board/freescale/imx93_evk/spl.c | 20 +++++++++++++++++++-
 1 file changed, 19 insertions(+), 1 deletion(-)

diff --git a/board/freescale/imx93_evk/spl.c b/board/freescale/imx93_evk/spl.c
index def3bab2509..ea42c056dbd 100644
--- a/board/freescale/imx93_evk/spl.c
+++ b/board/freescale/imx93_evk/spl.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0+
 /*
- * Copyright 2022 NXP
+ * Copyright 2022, 2024-2025 NXP
  */
 
 #include <common.h>
@@ -35,9 +35,27 @@
 #include <power/pca9450.h>
 #include <power/pf0900.h>
 #include <asm/arch/trdc.h>
+#include <serial.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#ifdef CONFIG_SPL_OS_BOOT
+int spl_start_uboot(void) {
+	static int last_call = -1;
+	if (last_call >= 0)
+		return last_call;
+
+	if (IS_ENABLED(CONFIG_SPL_SERIAL) && serial_tstc() && serial_getc()) {
+	/* break into full u boot if any key is pressed */
+		while(serial_tstc()) { /* eat up all the remaining characters */
+			serial_getc();
+		}
+		return last_call = 1;
+	}
+	return last_call = 0; // start the Kernel
+}
+#endif
+
 int spl_board_boot_device(enum boot_device boot_dev_spl)
 {
 #ifdef CONFIG_SPL_BOOTROM_SUPPORT
-- 
2.34.1

