From 4b04ab51ef1b29e725d9e420e9247ac34f577505 Mon Sep 17 00:00:00 2001
From: Elena Popa <elena.popa@nxp.com>
Date: Tue, 12 Aug 2025 17:03:27 +0300
Subject: [PATCH] imx8ulp: add falcon mode support

Upstream-Status: Inappropriate [board-specific Falcon Mode support]

Signed-off-by: Elena Popa <elena.popa@nxp.com>
---
 board/freescale/imx8ulp_evk/spl.c | 18 ++++++++++++++++++
 1 file changed, 18 insertions(+)

diff --git a/board/freescale/imx8ulp_evk/spl.c b/board/freescale/imx8ulp_evk/spl.c
index aa68afd69fa..25dffac118b 100644
--- a/board/freescale/imx8ulp_evk/spl.c
+++ b/board/freescale/imx8ulp_evk/spl.c
@@ -22,9 +22,27 @@
 #include <asm/mach-imx/ele_api.h>
 #include <asm/sections.h>
 #include <asm/mach-imx/boot_mode.h>
+#include <serial.h>
 
 DECLARE_GLOBAL_DATA_PTR;
 
+#ifdef CONFIG_SPL_OS_BOOT
+int spl_start_uboot(void) {
+	static int last_call = -1;
+	if (last_call >= 0)
+		return last_call;
+
+	if (IS_ENABLED(CONFIG_SPL_SERIAL) && serial_tstc() && serial_getc()) {
+	/* break into full u boot if any key is pressed */
+		while(serial_tstc()) { /* eat up all the remaining characters */
+			serial_getc();
+		}
+		return last_call = 1;
+	}
+	return last_call = 0; // start the Kernel
+}
+#endif
+
 void spl_dram_init(void)
 {
 	/* Reboot in dual boot setting no need to init ddr again */
-- 
2.34.1

