From 89d5dd025d1955f5d02b494f50a64a4b240b6cc6 Mon Sep 17 00:00:00 2001
From: Michal Gorlas <michal.gorlas@9elements.com>
Date: Thu, 20 Aug 2026 22:49:21 +0200
Subject: [PATCH] imx8mp: add falcon mode support

Signed-off-by: Michal Gorlas <michal.gorlas@9elements.com>
Upstream-Status: Pending
---
 board/nxp/imx8mp_evk/spl.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/board/nxp/imx8mp_evk/spl.c b/board/nxp/imx8mp_evk/spl.c
index 165a8cf585f..e5e8308a491 100644
--- a/board/nxp/imx8mp_evk/spl.c
+++ b/board/nxp/imx8mp_evk/spl.c
@@ -19,6 +19,25 @@
 #include <asm/arch/sys_proto.h>
 #include <asm/mach-imx/boot_mode.h>
 #include <asm/sections.h>
+#include <serial.h>
+
+
+#ifdef CONFIG_SPL_OS_BOOT
+int spl_start_uboot(void) {
+	static int last_call = -1;
+	if (last_call >= 0)
+		return last_call;
+
+	if (IS_ENABLED(CONFIG_SPL_SERIAL) && serial_tstc() && serial_getc()) {
+	/* break into full u boot if any key is pressed */
+		while(serial_tstc()) { /* eat up all the remaining characters */
+   		serial_getc();
+		}
+		return last_call = 1;
+	}
+ 	return last_call = 0; // start the Kernel
+ }
+ #endif
 
 int spl_board_boot_device(enum boot_device boot_dev_spl)
 {
-- 
2.47.3

